src/passes/Asyncify.cpp

```diff
@@ -1218,6 +1248,202 @@ struct AsyncifyFlow : public Pass {
   }
 };
 
+// Add catch block counters to verify that unwinding is not done from catch
+// block.
+struct AsyncifyAddCatchCounters : public Pass {
+  bool isFunctionParallel() override { return true; }
+
+  std::unique_ptr<Pass> create() override {
+    return std::make_unique<AsyncifyAddCatchCounters>();
+  }
+
+  void runOnFunction(Module* module_, Function* func) override {
+    class CountersBuilder : public Builder {
+    public:
+      CountersBuilder(Module& wasm) : Builder(wasm) {}
+      Expression* makeInc(int amount = 1) {
+        return makeGlobalSet(
+          ASYNCIFY_CATCH_COUNTER,
+          makeBinary(AddInt32,
+                     makeGlobalGet(ASYNCIFY_CATCH_COUNTER, Type::i32),
+                     makeConst(int32_t(amount))));
+      }
+      Expression* makeDec(int amount = 1) {
+        return makeGlobalSet(
+          ASYNCIFY_CATCH_COUNTER,
+          makeBinary(SubInt32,
+                     makeGlobalGet(ASYNCIFY_CATCH_COUNTER, Type::i32),
+                     makeConst(int32_t(amount))));
+      }
+    };
+
+    // with this walker we will handle those changes of counter:
+    // - entering top-level catch (= pop)           +1
+    // - entering nested catch (= pop)               0 (ignored)
+    //
+    // - return inside top-level/nested catch       -1
+    // - return outside top-level/nested catch       0 (ignored)
+    //
+    // - break target outside of top-level catch    -1
+    // - break target inside of top-level catch      0 (ignored)
+    // - break outside top-level/nested catch        0 (ignored)
+    //
+    // - exiting from top-level catch               -1
+    // - exiting from nested catch                   0 (ignored)
```

Comment:
Unfortunately, detecting an exit from a catch is not as simple as this. You would also need to handle `br_on` and `return_call`, off the top of my head.

We could in theory say this feature doesn't support GC or tail calls, but imagine this is linked with a module using those features, and then inlining happens, so it seems worrying.

How important is `asyncify-ignore-unwind-from-catch`? Without that, we could use the more static approach I mentioned before.

Reply:
I mean if needed devs can fix code to support something like ignore this asyncify_sleep

Reply:
Ok, I don't think anyone else mentioned a need for that option, so maybe it is best to remove it. Can maybe wait a bit to see if anyone on this PR responds.